[READY] improve window, application

lowne opened the pull request:
- added `hs.window.find()`, `hs.application.find()` (similar to #501, which is a prerequisite of this PR for `win:moveToScreen'dell'`) - `hs.appfinder` isn't needed anymore
- `hs.window.get()` and `hs.application.get()` are "strict" variants for `find`
- `hs.application:findWindow()` and `hs.application:getWindow()`
- added `hs.application.open()` (wraps the two launchOrFocus methods)
- `hs.application:title()->hs.application:name()` (users shouldn't have to suffer just because it derives from uielement)
- geometrified `hs.window` for cleaner code (#471, prerequisite), and to enable...
- `hs.window:move()` one-stop-shop for window movement; takes a point (moves by x,y), a rect (does setFrame) or a unit rect+screen (basically does :moveToScreen+:moveToUnit in one go) 
- unfnfied everything for correctness (there was a potential heisenbug in find_windows_in_direction), readability, maintainability and performance; in particular, **the occasional 3 to 5 seconds delay** when doing `hs.window.allWindows()` and derivatives (including stuff like #497) **seems now gone**; it used to happen quite often here, and I haven't seen it yet
- fixed some minor bugs, docs
- ensureIsInScreenBounds doesn't care about animations (fixes #478)
- also needs the windowfilter->window.filter PR (#513) ~~which apparently I haven't opened yet :/~~

lowne commented:
I think there's a good chance that filtering non-gui apps in `window.allWindows()` (the "major speedup" commit here) will help wrt #277, #175 and the like.

cmsj commented:
@lowne What's the reasoning for the app:kind() test being >0? Shouldn't apps with a kind() of 0 be able to own windows?

lowne commented:
@cmsj wtf, yes... I explicitly remember thinking "kind must be >=0" (as in windowfilter) - it slipped through somehow. That was an important one to catch! Will fix.

lowne commented:
Fixed; and added `hs.window._timed_addWindows()` (to be run from the console) to help in diagnosing slowness problems such as in the previously mentioned issues.

cmsj commented:
@lowne also, it looks like all of the docstrings in hs.appfinder are being rewritten by this, to not be docstrings anymore? (i.e. --- replaced with --)

cmsj commented on extensions/application/init.lua:
```diff
@@ -19,7 +21,11 @@ application.watcher = require "hs.application.watcher"
 ---  * A table containing zero or more hs.window objects
 function application:visibleWindows()
-  return fnutils.filter(self:allWindows(), window.isVisible)
+  --  return moses.filter(self:allWindows(), window.isVisible)
```
Could we remove the moses reference(s) please?

cmsj commented on extensions/application/init.lua:
```diff
@@ -48,4 +48,145 @@ function application:activate(_allWindows)
+---  * the `hs.application` object for the launched or activated application; `nil` if not found
+---
+--- Notes:
+---  * the `wait` parameter will *halt the entire script* in order to return the application object "synchronously"; only use it if you
```
Rather than "halt the entire script", maybe this should say "block all Hammerspoon activity", since timer.usleep() will block the entire application.

cmsj commented on extensions/window/init.lua:
```diff
@@ -8,14 +8,12 @@
 local window = require "hs.window.internal"
 local application = require "hs.application.internal"
-local fnutils = require "hs.fnutils"
+--local moses = require "hs.moses"
```
Ditto removing moses.

cmsj commented on extensions/window/init.lua:
```diff
@@ -25,30 +23,131 @@ local atan,cos,abs = math.atan,math.cos,math.abs
---- hs.window.allWindows() -> win[]
---- Function
+--- hs.window.allWindows() -> list of hs.window objects
+--- Constructor
```
I don't think this is a constructor - they should typically be returning an object that has a metatable.

cmsj commented on extensions/window/init.lua:
```diff
@@ -25,30 +23,131 @@ local atan,cos,abs = math.atan,math.cos,math.abs
+end
+
+--- hs.window.visibleWindows() -> list of hs.window objects
+--- Constructor
```
Ditto constructor

cmsj commented on extensions/window/init.lua:
```diff
@@ -25,30 +23,131 @@ local atan,cos,abs = math.atan,math.cos,math.abs
+end
+
+--- hs.window.orderedWindows() -> list of hs.window objects
+--- Constructor
```
Ditto constructor

cmsj commented on extensions/window/init.lua:
```diff
@@ -25,30 +23,131 @@ local atan,cos,abs = math.atan,math.cos,math.abs
+window.windowForID=window.get
+
+--- hs.window.find(hint) -> hs.window object(s)
+--- Constructor
```
Ditto constructor

cmsj commented on extensions/window/init.lua:
```diff
@@ -328,71 +393,66 @@ end
+  end
+  return maxs
+end
+--[[ -- moses version
```
ditto moses

cmsj commented:
This should also help with #233, or at least allow us to apply similar fixes until that is solved.

lowne commented:
@cmsj yes, the idea being that application/window.find/get obviate the need for a separate module (but staying for now for backward compatibility) - I changed the appfinder functions to be thin wrappers.

cmsj commented:
@lowne I think I'd prefer to keep the docs around for a while, with a note at the top saying that the module is deprecated. It's pretty widely used, afaik.

lowne commented:
@cmsj done
